### ISSUE TITLE
Discuss: a unified model for cells / cell state for Engine and Host/Contexts

### ISSUE THREAD

**Issue by obuchtala** [stencila/engine#2]
ATM there are two different notions of a cell state, one used by the Engine to track the state and dependencies of cells, and one used by the Host/Contexts to perform evaluations.

- the Engine's internal cell model can be looked at here: https://github.com/stencila/engine/blob/65a5494c268dd808e443c4c0ecd941e1d9b8973f/src/Cell.js

- the Host's notion of cells is documented here: https://github.com/stencila/website/blob/c704fedbdf72da821da3a737123a254cedffa63e/src/specs/cell.yaml

These are two different angles with a different set of requirements.
We want to identify where the common denominator is, which data needs to be exposed to a programmer, and for which purpose.

WIP: to be continued.

**Comment by obuchtala** [comment]
> https://github.com/stencila/website/blob/c704fedbdf72da821da3a737123a254cedffa63e/src/specs/cell.yaml#L4

What is this used for?

**Comment by obuchtala** [comment]
> https://github.com/stencila/website/blob/c704fedbdf72da821da3a737123a254cedffa63e/src/specs/cell.yaml#L22

To indicate this is a boolean, I would prefer `isExpr`, or something similar.

**Comment by obuchtala** [comment]
> https://github.com/stencila/website/blob/c704fedbdf72da821da3a737123a254cedffa63e/src/specs/cell.yaml#L26

How is this used exactly?

**Comment by obuchtala** [comment]
> https://github.com/stencila/website/blob/c704fedbdf72da821da3a737123a254cedffa63e/src/specs/cell.yaml#L30

In the Engine this has a different meaning: it is more the symbol, rather than the value, i.e. to be able to determine dependencies it is important to know how it is formulated in the source. OTOH for a Context, only the actual value is necessary.

**Comment by obuchtala** [comment]
> https://github.com/stencila/website/blob/c704fedbdf72da821da3a737123a254cedffa63e/src/specs/cell.yaml#L40

Similar to the inputs, the Engine needs the symbol information, not the value.
On the other hand, there is currently only a single output supported per cell.

**Comment by obuchtala** [comment]
On the other side (Engine) there are a bunch of properties which are not needed for execution:
- `status`:  https://github.com/stencila/engine/blob/65a5494c268dd808e443c4c0ecd941e1d9b8973f/src/Cell.js#L42
- `errors`: https://github.com/stencila/engine/blob/65a5494c268dd808e443c4c0ecd941e1d9b8973f/src/Cell.js#L48
- `value` has been distinct from the symbolic representation of output because a cell can have a value, while not producing an output. https://github.com/stencila/engine/blob/65a5494c268dd808e443c4c0ecd941e1d9b8973f/src/Cell.js#L50
- internal stuff that would probably not be part of a high-level cell API, until we want to start to think about interruptible execution, i.e. dump engine state and continue later

**Comment by obuchtala** [comment]
@nokome just a quick scan over the two 'schemas' identifying the differences. Pls add your thoughts :) Thanks!

**Comment by nokome** [comment]
Thanks @oliver---- : I am going to move the specs into a separate repo today and hopefully find the time to answer some of your questions above over there, in the YAML.

**Comment by obuchtala** [comment]
Fine :) You have some time, as I am leaving for sleep now. Talk later :)

**Comment by nokome** [comment]
I've moved these cell specs to https://github.com/stencila/specs/blob/master/src/Cell.yaml I've added some more description and comments there too.

FWIW, there are also examples of the pre and post execute cells in a more digestible form in the Host API spec
https://stencila.github.io/specs/host.html#/paths/~1instances~1{id}~1execute/put

**Comment by obuchtala** [comment]
@nokome I am working on using your suggested format in the engine.
I fear I don't like using arrays for inputs and outputs. This is in the way all the time in a Context and in the Engine, and also suggests that the order is of relevance. I'd suggest to use Maps instead.
I don't think it is necessary to have exactly the same native structures for execution and serialisation. And an array of entries is the common serialisation of a map.
What do you think?

**Comment by obuchtala** [comment]
And re the term 'expr': I guess with `global` it is better because it is an adjective, thus a better fit for a boolean flag. Maybe we could find an adjective that describes what the `expr` flag means?

**Comment by obuchtala** [comment]
Thinking about the returned value. I don't like how implicit this is currently. I.e. in the code there you can find a lot of `outputs[0]`.
To make this more explicit, we could introduce a new field `returns` (or something alike)
We could use a special name for an unnamed output.

**Comment by nokome** [comment]
> I don't think it is necessary to have exactly the same native structures for execution and serialisation.

Agreed. This seems like an important principle that we should use to bridge between a serialisation/interchange structure for a cell (which is flexible enough that it doesn't need to change often e.g. future proofed to allow for things like multiple names outputs from a cell) and implementation structures which handle cells (e.g. `Engine`, `JavascriptContext`)

The reason I made `inputs` an array was precisely to allow order to be of relevance and in the future to allow cells to be treated as functions (e.g. a block data-flow type visual programming UI, where the user connects several inputs to a cell without naming them). But maybe that is the wrong approach and instead of trying to create a one-size fits all we should focus on separate `Cell` and `IIFE`-like structures which are optimised for each use case and convert between them.

In relation to output/s name/value...

> I don't like how implicit this is currently. I.e. in the code there you can find a lot of outputs[0]. To make this more explicit, we could introduce a new field returns (or something alike)

Agreed that all the `outputs[0] ` are smelly. But instead of adding a `returns` field to the serialisation format, the implementation could have a `Cell` class with `import` and `export` methods which map a structure which is optimised for implementation, to one that is optimised for serialisation.

Nonetheless, it would be nice to have a serialisation and implementation structures (across languages) that were as close as possible and using objects/maps, instead of arrays for both `inputs` and `outputs` may be a better way to go.

Mmm, that was a bit of a ramble :) maybe we should chat again
